### ISSUE TITLE
ERROR TypeError: this.client_.unaryCall is not a function

### ISSUE THREAD

**Issue by AlexDaSoul** [grpc/grpc-web#477]
https://github.com/grpc/grpc-web/blob/master/packages/grpc-web/index.d.ts

**Comment by stanley-cheung** [comment]
I think I know what this is. Can you give me a quick steps to reproduce please? Thanks.

**Comment by AlexDaSoul** [comment]
Just generate files with last version. I use import_style=commonjs+dts

**Comment by stanley-cheung** [comment]
Sorry I still can't quite reproduce this. I generated using import_style=commonjs+dts. I ran `tsc` but got no error. I really need to see that error myself.

**Comment by AlexDaSoul** [comment]
Method unaryCall used in your generated files?

**Comment by AlexDaSoul** [comment]
Now i use revision 73358722d8223955cc6d6d5be0c8c483988a0f0d and and everything is ok

```
RUN cd /github/grpc-web && \
  git checkout -b old-fix 73358722d8223955cc6d6d5be0c8c483988a0f0d && \
  ./scripts/init_submodules.sh
```

In newest versions method unaryCall is not defined into node_modules/grpc-web, but used in generated files

`/**
 * @param {!proto.sdex.account.UpdateAddressRequest} request The
 *     request proto
 * @param {?Object<string, string>} metadata User defined
 *     call metadata
 * @return {!Promise<!proto.sdex.account.UpdateAddressResponse>}
 *     A native promise that resolves to the response
 */
proto.sdex.account.AddressManagerServicePromiseClient.prototype.update = function(
    request,
    metadata
) {
    return this.client_.unaryCall(
        this.hostname_ + '/sdex.account.AddressManagerService/Update',
        request,
        metadata || {},
        methodInfo_AddressManagerService_Update
    );
};`

**Comment by stanley-cheung** [comment]
Yea I see that `.unaryCall` is not defined in the latest node_modules/grpc-web yet, but how do I trigger that error message of "TypeError: this.client_.unaryCall is not a function"?

**Comment by AlexDaSoul** [comment]
You need to call the service and method api that uses unaryCall. Forexample i call `AddressManagerServicePromiseClient.update({...}, meta);`

**Comment by foisonocean** [comment]
Any updates? I have the same problem, and is there any solutions for that?

**Comment by AlexDaSoul** [comment]
My solution was to take a certain revision. But this is a bad solution, although the only possible one in this situation. Therefore, I want to ask the distinguished author to be more responsible to versioning :)

**Comment by stanley-cheung** [comment]
I finally understand what this is. 

This is actually *not* about the `AbstractClientBase.unaryCall` function not being defined in `index.d.ts`. What happened is this:

1. Compile the latest `protoc-gen-grpc-web` plugin from the latest of `master`.
2. Write a client that uses the `xxxPromiseClient`
3. When compiling the client (either JS or TS is fine), at the `npm install` step, notice that this is pulling the `grpc-web` npm package from the *released* version of 1.0.3, not from `master`.

That's how the bug should be reproduced. This is just a mismatch runtime problem - you are trying to call an RPC function from your generated `xxxPromiseClient` (from latest `master`), which does call down to the newly introduced (but not released) `.unaryCall` library function. But, the `.unaryCall` library function has not been released in the `grpc-web` npm package runtime yet. 

So there are 2 ways to solve this:
1. Make sure you are either using both the *release* version of the `protoc-gen-grpc-web` plugin and the `grpc-web` runtime package, OR
2. if you really want to test out the latest `protoc-gen-grpc-web` plugin from `master`, you need to also compile the `grpc-web` npm package from `master`, locally and then `npm link` them.

Basically, the `protoc-gen-grpc-web` plugin and the `grpc-web` npm package runtime need to be in lock-step.

To really drill down on how exactly to do 2. above
- Here's how to compile the `grpc-web` npm package from `master`, locally, https://github.com/grpc/grpc-web/blob/master/net/grpc/gateway/docker/prereqs/Dockerfile#L29-L31
- Then when you compile your project, you need to run `npm link grpc-web`, like this: https://github.com/grpc/grpc-web/blob/master/net/grpc/gateway/docker/ts_client/Dockerfile#L27

Actually that's how we run our tests -  we have one set of tests which downloads the released version of the plugin and use the released version of the npm package. For another set of tests, we compile both the plugin and the npm package from `master`. 

Sorry for the long wall of text. Hope this clarifies the situation!

**Comment by stanley-cheung** [comment]
On the other hand, the `.unaryCall` function type def missing from our package's `index.d.ts` will indeed become a problem down the road. But that will manifest itself in a different way.

It gets a bit complicated to explain. In short, `import_style=commonjs+dts` actually will not be a problem because the type checking sort of stops at the generated client stub classes level. It will become a problem if the client is generated with the full `import_style=typescript`. However, `import_style=typescript` does not generate the `xxxPromiseClient` for now so there's currently no way to trigger trigger this error at the TypeScript level 😉 

I will definitely follow up on adding `.unaryCall` to `index.d.ts` but that will mean that we need some code path that will trigger that.

**Comment by AlexDaSoul** [comment]
Thanks! But I can't check it because new version doesn't include my enums :(

**Comment by stanley-cheung** [comment]
We have just released version `1.0.4`. Please try again. This error should be gone if you are using the released version of the `protoc-gen-grpc-web` plugin with the `1.0.4` version of the npm package.

**Comment by AlexDaSoul** [comment]
It's work
Thank you
